### PR TITLE
added l1 penlaty

### PR DESCRIPTION
**Pull request by yorkerlin**
I will update the document soon.

### CODE REVIEW

File: src/shogun/optimization/L1Penalty.h [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+#include <shogun/optimization/Penalty.h>
+#include <shogun/lib/config.h>
+#include <shogun/mathematics/Math.h>
+namespace shogun
+{
+/** @brief The is the base class for L1 penalty/regularization within the FirstOrderMinimizer framework.
+ *
+ * For L1 penalty, \f$L1(w)\f$
+ * \f[
+ * L1(w)=|w|
```
Comment (lisitsyn) [inline src/shogun/optimization/L1Penalty.h]: ```\| w \|_1 = \sum_i \| w_i \|```

File: src/shogun/optimization/L1Penalty.h [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+	virtual float64_t get_penalty_gradient(float64_t variable,
+		float64_t gradient_of_variable)=0;
+
+	virtual float64_t get_sparse_variable(float64_t variable, float64_t penalty_delta)=0;
+
+
+	/** Does the penalty enforce the sparsity constraint ?
+	 * @return true if it enforces the constraint
+	 */
+	virtual bool enforce_sparsity() {return true;}
```
Comment (lisitsyn) [inline src/shogun/optimization/L1Penalty.h]: This looks a bit strange to me.. What is it for?

Comment (yorkerlin) [inline src/shogun/optimization/L1Penalty.h]: @lisitsyn 
For smooth penalty (eg, L2), SGD methods will do `compute sample gradient` and then update the gradient for `smooth penalty` and then update the variable.

For non-smooth penalty (eg, L1), the produce mentioned does not give the sparse variable.   
For L1, we have to `compute sample gradient` , then update the variable and finally use the penalty  (`get_sparse_variable`) to generate sparse variable.

We have to distinguish these two produces.

**Comment by yorkerlin** [comment]
@lisitsyn 
I will update this PR after https://github.com/shogun-toolbox/shogun/pull/2910 is merged

File: src/shogun/optimization/L1Penalty.h (hunk 1 above)
Comment (yorkerlin) [inline src/shogun/optimization/L1Penalty.h]: thanks

File: src/shogun/optimization/L1Penalty.h (hunk 2 above)
Comment (yorkerlin) [inline src/shogun/optimization/L1Penalty.h]: it is for `L1 penalty` and `Elastic Net ( a combination of L1 and L2)`

**Comment by yorkerlin** [comment]
I close this PR.
Please see the new PR (https://github.com/shogun-toolbox/shogun/pull/2913)

**Pull request closed without merging**